litgpt/model.py

```diff
@@ -410,17 +437,67 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
 
 def build_rope_cache(
-    seq_len: int, n_elem: int, device: Optional[torch.device] = None, base: int = 10000, condense_ratio: int = 1
+    seq_len: int,
+    n_elem: int,
+    device: Optional[torch.device] = None,
+    base: int = 10000,
+    condense_ratio: int = 1,
+    extra_config: Optional[dict] = None,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
     """Enhanced Transformer with Rotary Position Embedding.
 
     Derived from: https://github.com/labmlai/annotated_deep_learning_paper_implementations/blob/master/labml_nn/
     transformers/rope/__init__.py. MIT License:
     https://github.com/labmlai/annotated_deep_learning_paper_implementations/blob/master/license.
+
+    Args:
+        seq_len (int): Sequence length.
+        n_elem (int): Number of elements (head dimension).
+        device (torch.device, optional): Device for tensor allocations.
+        base (int, optional): Base for computing inverse frequencies.
+        condense_ratio (int, optional): Ratio to condense the position indices.
+        extra_config (dict, optional): Configuration parameters for frequency adjustments (used by Llama 3.1 and 3.2)
+
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor]: Cosine and sine caches for RoPE.
     """
-    # $\Theta = {\theta_i = 10000^{\frac{2(i-1)}{d}}, i \in [1, 2, ..., \frac{d}{2}]}$
+
+    # $\Theta = {\theta_i = 10000^{\frac{2(i-1)}{d}}, i \in [1, 2, ..., \frac{d}{2}]}$	    assert n_elem % 2 == 0, "n_elem (head dimension) must be even"
     theta = 1.0 / (base ** (torch.arange(0, n_elem, 2, device=device).float() / n_elem))
 
+    if extra_config is not None:
+        # Extract configuration parameters
+        orig_context_len = extra_config["original_max_seq_len"]
+        factor = extra_config["factor"]
+        low_freq_factor = extra_config["low_freq_factor"]
+        high_freq_factor = extra_config["high_freq_factor"]
+
+        # Compute wavelength thresholds
+        low_freq_wavelen = orig_context_len / low_freq_factor
+        high_freq_wavelen = orig_context_len / high_freq_factor
+
+        # Compute wavelengths corresponding to the inverse frequencies
+        wavelen = 2 * torch.pi / theta
+
+        # Initialize adjusted inverse frequencies
+        adjusted_theta = theta.clone()
+
+        # Low Frequency Region: wavelen > low_freq_wavelen
+        mask_low_freq = wavelen > low_freq_wavelen
+        adjusted_theta[mask_low_freq] = theta[mask_low_freq] / factor
+
+        # Medium Frequency Region: high_freq_wavelen ≤ wavelen ≤ low_freq_wavelen
```

Comment:
```suggestion
        # Medium Frequency Region: low_freq_wavelen ≤ wavelen ≤ high_freq_wavelen
```

Just assuming by the name - medium frequency region.
Which means that the mask also might be incorrect.

Reply:
Thanks for the comment! I always appreciate a second pair of eyes! Regarding this particular suggestion:

> # Medium Frequency Region: low_freq_wavelen ≤ wavelen ≤ high_freq_wavelen

I think this is impossible.

Let's assume we have
- `high_freq_wavelen=10`
- `low_freq_wavelen=100`

How can this condition be met?

`100 ≤ wavelen ≤ 10` would not be possible, or am I misunderstanding?

Reply:
Oops.
Sometimes I can surprise even myself with my "intelligence".
Who would have thought that in order to have a higher frequency you need a shorter wavelength. 😂

Reply:
Ohhhh, yeah, well, it's one of these confusing things in physics. But phew! And I was worried that the implementation was wrong haha (but there should be a unit test with the transformer library)